k8s/src/main/scala/io/buoyant/k8s/EndpointsNamer.scala

```diff
@@ -52,60 +57,75 @@ class EndpointsNamer(idPrefix: Path, mkApi: String => NsApi) extends Namer {
+      val state = Var[Activity.State[T]](Activity.Pending)
+      _retryToActivity(backoff, state)(go)
+      Activity(state)
+    }
+
+    private[this] def _retryToActivity[T](
+      remainingBackoff: Stream[Duration],
+      state: Var[Activity.State[T]] with Updatable[Activity.State[T]] = Var[Activity.State[T]](Activity.Pending)
+    )(go: => Future[T]): Unit = {
+      val _ = go.respond {
+        case Return(t) =>
+          state() = Activity.Ok(t)
+        case Throw(e) =>
+          state() = Activity.Failed(e)
+          remainingBackoff match {
+            case delay #:: rest =>
+              Future.sleep(delay).onSuccess { _ => _retryToActivity(rest, state)(go) }
+            case Stream.Empty =>
+          }
+      }
+    }
 
     def get(name: String): NsCache = synchronized {
       caches.get(name) match {
         case Some(ns) => ns
         case None =>
-          val ns = mkNs(name)
+          val ns = new NsCache(name)
+          val closable = retryToActivity { watch(name, ns) }
+          _watches += (name -> closable)
           caches += (name -> ns)
           ns
       }
     }
 
-    private[this] def mkNs(name: String): NsCache = {
-      val nsCache = new NsCache(name)
-      _watches = _watches + (name -> watch(name, nsCache))
-      nsCache
-    }
-
-    // XXX once a namespace is watched, it is watched forever.  also
-    // theres probably some poor error behavior.
-    private[this] var _watches = Map.empty[String, Closable]
-    def watches = _watches
-
-    private[this] def watch(namespace: String, services: NsCache): Closable = {
+    private[this] def watch(namespace: String, services: NsCache): Future[Closable] = {
       val ns = mkApi(namespace)
       val endpointsApi = ns.endpoints
-      val close = new AtomicReference(Closable.nop)
 
       Trace.letClear {
-        val init = {
-          log.debug("k8s initializing %s", namespace)
-          val endpoints = endpointsApi.get()
-
-          close.set(Closable.make { _ =>
-            endpoints.raise(Closed)
-            Future.Unit
-          })
-
-          endpoints.onSuccess { list =>
-            services.initialize(list)
-          }.onFailure { e =>
-            log.error(e, "k8s failed to list endpoints")
-          }
-        }
-
-        init.foreach { init =>
+        log.debug("k8s initializing %s", namespace)
+        endpointsApi.get().map { list =>
+          services.initialize(list)
           val (updates, closable) = endpointsApi.watch(
-            resourceVersion = init.metadata.flatMap(_.resourceVersion)
+            resourceVersion = list.metadata.flatMap(_.resourceVersion)
           )
-          close.set(closable)
-          updates.foreach(services.update)
+          // fire-and-forget this traversal over an AsyncStream that updates the services state
+          val _ = updates.foreach(services.update)
+          closable
+        }.onFailure { e =>
+          log.error(e, "k8s failed to list endpoints")
```

Comment:
tioli: The prior `AtomicReference` approach before made it possible to cancel the request if its closed before the request is satsified. Practically, it probably doesn't matter, but it seems more correct to do so....